design/control-data-plane-separation/design.md

```diff
@@ -23,377 +23,118 @@ architecture does not allow the control plane and data plane to scale independen
-the [Command Message](#command-message) section.
-
-Commands can contain payloads, but they are mostly small (a few kb at most) and should not transport large amounts of
-data. The message size is significant with gRPC because gRPC has a message size limit. For larger payloads, like the
-nginx configuration, the `Download` and `Upload` RPCs will be used in conjunction with commands.
-
-The control plane will use the `Download` channel to push the configuration to the agent. See
-the [Configuration Download](#configuration-download) section for details.
+not an issue. The gRPC runtime will handle the connection establishment and management. If an error occurs or the stream or connection is dropped, the connection must be reestablished.
 
-The agent uses the `Upload` command to upload large payloads, such as the nginx configuration, to the control plane. For
-our use case, we only need the ability to send the nginx configuration from the control plane to the agent. We do not
-need the agent to upload its configuration to the control plane. Ideally, we would like to disable this functionality in
-the agent.
+#### Further Requirements and Implementation Details
 
-For the full commander service definition, see
-this [file](https://github.com/nginx/agent/blob/main/sdk/proto/command_svc.proto).
+- The control plane will need to run a gRPC server for the agent to connect to.
+- When an agent connects to the control plane, the payload _should_ contain the hostname (pod name) and nginx instanceID of the registering agent. This can be used to keep track of all agents/pods that are connected.
+  - We need to be able to link an agent connection with a subscription. These are two different gRPC calls. `Subscribe` is where we actually send an nginx config to an agent. We need to ensure that we are sending the correct nginx config to the correct agent. Ideally we use metadata from the agent connection (maybe hostname, maybe we need middleware to extract token/uuid from grpc connection headers) and store that in the context. The context is then used in the `Subscribe` call, where we can extract the agent identifier and send the proper nginx config for that identifier.
 
-#### Command Message
+  Process: agent `Connects` to NGF. We get its identifier and pod name, add the identifier(s) to a context cache, track that connection, and create a subscription channel for it. Agent then `Subscribes`. The context passed in allows us to use the identifier to grab the proper subscription channel and listen on it. This channel will receive a `ConfigApplyRequest` when we have a new nginx config to write.
 
-Command Messages have the following structure:
+- If a single nginx deployment is scaled, we should ensure that all instances for that deployment receive the same config (and aren't treated as "different").
+- Each graph that the control plane builds internally should be directly tied to an nginx deployment.
```

Comment:
Do we need one graph per deployment? Or can we keep the graph structure we have and just add more Gateways to it? Each Gateway "node" corresponds to an nginx deployment.

Reply:
Yeah, in my head it really is a single "graph", but with multiple Gateway trees inside of it.